### PR TITLE
fix(downsampler): update partitionkey dataschema to downsampled schema

### PR DESCRIPTION
**Pull request by sherali42**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?

**Current behavior :**
- Downsampled Chunks are persisted with partition keys having raw data-schema, causing failures in downsampled queries.
- Partition Key index updates are also persisted with raw data-schema

**New behavior :**
- PartKey data-schema is updated with downsampled schema for both downsampled chunks and PartKey chunks

### CODE REVIEW

File: spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala [hunk 1]
```diff
@@ -362,5 +366,13 @@ object BatchDownsampler extends StrictLogging with Instance {
       blacklist.forall(w => !w.forall(pkPairs.contains))
     }
   }
+  /**
+    * This method is used for updating Partkey's schema to downsampled schema as gauge has different
+    * downsampled schema.
+    * schemaId is stored at offset+4.
+    */
+  private[downsampler] def updateDataSchema(partKeyBase: Array[Byte], partKeyOffset: Long, schemaId: Short): Unit = {
+    UnsafeUtils.setShort(partKeyBase, partKeyOffset + 4, schemaId)
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: We need better abstraction here than this. This code will be affected if encoding of part keys change. Lets use a utility API that is closer to the part key encoding class.

File: spark-jobs/src/test/scala/filodb/downsampler/DownsamplerMainSpec.scala [hunk 2]
```diff
@@ -553,4 +565,28 @@ class DownsamplerMainSpec extends FunSpec with Matchers with BeforeAndAfterAll w
...
+      1000, InProcessPlanDispatcher, BatchDownsampler.rawDatasetRef, 0, queryFilters, AllChunkScan,
+      colName = Option("sum"))
+    val queryConfig = new QueryConfig(DownsamplerSettings.filodbConfig.getConfig("query"))
+    val queryScheduler = Scheduler.fixedPool(s"$QuerySchedName", 3)
+    val res = exec.execute(downsampleTSStore, queryConfig)(queryScheduler, 1 minute)
+      .runAsync(queryScheduler).futureValue.asInstanceOf[QueryResult]
+    queryScheduler.shutdown()
+    res.result.size shouldEqual 1
+    res.result.head.rows.map(r => (r.getLong(0), r.getDouble(1))).toList shouldEqual
+      List((1574273042000L, 112.0))
```
Comment (vishramachandran) [inline spark-jobs/src/test/scala/filodb/downsampler/DownsamplerMainSpec.scala]: was only one sample returned ? Is this right ?

Comment (sherali42) [reply spark-jobs/src/test/scala/filodb/downsampler/DownsamplerMainSpec.scala]: @vishramachandran confirmed with Vish, this is correct

File: spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala (hunk 1 above)
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: moved this method to RecordSchema

File: spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala [hunk 3]
```diff
@@ -362,5 +366,13 @@ object BatchDownsampler extends StrictLogging with Instance {
       blacklist.forall(w => !w.forall(pkPairs.contains))
     }
   }
+  /**
+    * This method is used for updating Partkey's schema to downsampled schema as gauge has different
+    * downsampled schema.
+    * schemaId is stored at offset+4.
+    */
+  private[downsampler] def updateDataSchema(partKeyBase: Array[Byte], partKeyOffset: Long, schemaId: Short): Unit = {
```
[the remark this answers is not in the document]
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: removed this and add a method in RecordSchema

File: spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala [hunk 4]
```diff
@@ -193,6 +193,10 @@ object BatchDownsampler extends StrictLogging with Instance {
         val bufferPool = offHeapMem.bufferPools(rawPartSchema.downsample.get.schemaHash)
         val downsamplers = chunkDownsamplersByRawSchemaId(rawSchemaId)
         val periodMarker = downsamplePeriodMarkersByRawSchemaId(rawSchemaId)
+
+        updateDataSchema(rawReadablePart.partKeyBase, rawReadablePart.partKeyOffset,
```
[the remark this answers is not in the document]
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: updated the code

File: spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  def downsampledPk(partKey: Array[Byte], dsSchema: Schema, offHeapMem: MemFactory): Array[Byte] = {
+    val pkStrPairs = schemas.part.binSchema.toStringPairs(partKey, UnsafeUtils.arayOffset)
+    downsampledPk(pkStrPairs, dsSchema, offHeapMem)
+  }
+
+  /**
+    * create downsampled Partkey from the source/raw Partkey
+    */
+  def downsampledPk(pkStrPairs: Seq[(String, String)], dsSchema: Schema, offHeapMem: MemFactory): Array[Byte] = {
+    val dsRecordBuilder = new RecordBuilder(offHeapMem)
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala]: Lets reuse RecordBuilder

File: spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala [hunk 6]
```diff
@@ -30,15 +32,30 @@ object DSIndexJob extends StrictLogging with Instance {
...
+  private val dsSchemas = settings.rawSchemaNames.map { s => schemas.schemas(s).downsample.get}
+
+
+  // FIXME * 4 exists to workaround an issue where we see under-allocation for metaspan due to
+  // possible mis-calculation of max block meta size.
+  private val maxMetaSize = dsSchemas.map(_.data.blockMetaSize).max * 4
+
+  private val offHeapMem = new OffHeapMemory(dsSchemas,
+    kamonTags, maxMetaSize, settings.downsampleStoreConfig)
+
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala]: Use Memfactory.onHeap

File: spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    val metric = pkStrPairs.find(_._1 == dsSchema.options.metricColumn).get._2
+    val map: mutable.HashMap[ZeroCopyUTF8String, ZeroCopyUTF8String] = mutable.HashMap.empty
+    pkStrPairs.filter(_._1 != dsSchema.options.metricColumn).foreach(a => map.put(a._1.utf8, a._2.utf8))
+    val partKeyOffset = dsRecordBuilder.partKeyFromObjects(dsSchema, metric, map.toMap)
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala]: We are assuming fixed schema here. Not ideal to hard-code schema.

File: core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala [hunk 8]
```diff
@@ -462,6 +462,30 @@ object RecordSchema {
...
+  /**
+    * Build a partkey from the source partkey and change the downsample schema.
+    * Useful during downsampling as dataschema may differ.
+    *
+    * @throws java.util.NoSuchElementException if there is not downsample schema
+    */
+  final def buildDSPartKey(pkByte: Array[Byte], schemas: Schemas): Array[Byte] = {
+    val dsPkeyByte = pkByte.clone
+    val dsSchemaId = schemas(schemaID(pkByte)).downsample.get
+    UnsafeUtils.setShort(dsPkeyByte, UnsafeUtils.arayOffset + 4, dsSchemaId.schemaHash.toShort)
```
Comment (vishramachandran) [inline core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala]: Call `updateDSSchema` from here

File: spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala [hunk 9]
```diff
@@ -30,15 +31,27 @@ object DSIndexJob extends StrictLogging with Instance {
...
+  /**
+    * Downsample Schemas
+    */
+  private val dsSchemas = settings.rawSchemaNames.map { s => schemas.schemas(s).downsample.get}
+
+
+  // FIXME * 4 exists to workaround an issue where we see under-allocation for metaspan due to
+  // possible mis-calculation of max block meta size.
+  private val maxMetaSize = dsSchemas.map(_.data.blockMetaSize).max * 4
+
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala]: Can remove all of the above lines

File: core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala [hunk 10]
```diff
@@ -462,6 +462,30 @@ object RecordSchema {
...
+    UnsafeUtils.setShort(UnsafeUtils.ZeroPointer, addr + 4, dsSchemaId.schemaHash.toShort)
+  }
+
+  /**
+    * Build a partkey from the source partkey and change the downsample schema.
+    * Useful during downsampling as dataschema may differ.
+    *
+    * @throws java.util.NoSuchElementException if there is not downsample schema
+    */
+  final def buildDSPartKey(pkByte: Array[Byte], schemas: Schemas): Array[Byte] = {
```
Comment (vishramachandran) [inline core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala]: Can we expand `DS` to `Downsample` in both method names?

File: spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala [hunk 11]
```diff
@@ -193,10 +194,13 @@ object BatchDownsampler extends StrictLogging with Instance {
         val bufferPool = offHeapMem.bufferPools(rawPartSchema.downsample.get.schemaHash)
         val downsamplers = chunkDownsamplersByRawSchemaId(rawSchemaId)
         val periodMarker = downsamplePeriodMarkersByRawSchemaId(rawSchemaId)
+
         val (_, partKeyPtr, _) = BinaryRegionLarge.allocateAndCopy(rawReadablePart.partKeyBase,
                                                    rawReadablePart.partKeyOffset,
                                                    offHeapMem.nativeMemoryManager)
 
+        RecordSchema.updateDSSchema(partKeyPtr, schemas)
```
Comment (vishramachandran) [inline spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: Call only if downsampleSchema != rawSchema ?

Comment (vishramachandran) [reply spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: Also add comment on why this is needed

Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/BatchDownsampler.scala]: fixed

File: core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala (hunk 10 above)
Comment (sherali42) [reply core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala]: fixed

File: spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala (hunk 9 above)
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala]: fixed

File: core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala (hunk 8 above)
Comment (sherali42) [reply core/src/main/scala/filodb.core/binaryrecord2/RecordSchema.scala]: fixed

File: spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala (hunk 7 above)
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala]: changed the code

File: spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala (hunk 6 above)
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/index/DSIndexJob.scala]: refactored the code

File: spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala (hunk 5 above)
Comment (sherali42) [reply spark-jobs/src/main/scala/filodb/downsampler/DownsamplerUtil.scala]: modified the code to use different strategy

**Comment by velvia** [comment]
@sherali42   I thought the agreed plan was to move the mutable code to RecordBuilder, where you can create a new part key from an old one and change the schema.  Recordbuilder is designed to handle changes and mutations, whereas RecordSchema is used in places to read only.  I think maintaining this separation is really important for safe code.

File: prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala [hunk 12]
```diff
@@ -21,6 +21,8 @@ trait Functions extends Base with Operators with Vectors {
       throw new IllegalArgumentException(s"Invalid function name [$name]")
     }
 
+
```
Comment (velvia) [inline prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala]: No need for extra blank lines

File: query/src/main/scala/filodb/query/PlanEnums.scala [hunk 13]
```diff
@@ -55,6 +55,8 @@ object InstantFunctionId extends Enum[InstantFunctionId] {
 
 sealed abstract class RangeFunctionId(override val entryName: String) extends EnumEntry
 
+
```
Comment (velvia) [inline query/src/main/scala/filodb/query/PlanEnums.scala]: Here too

File: prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala (hunk 12 above)
Comment (sherali42) [reply prometheus/src/main/scala/filodb/prometheus/ast/Functions.scala]: ack, fixed it

File: query/src/main/scala/filodb/query/PlanEnums.scala (hunk 13 above)
Comment (sherali42) [reply query/src/main/scala/filodb/query/PlanEnums.scala]: ack, fixed it

**Pull request merged**